tests/end_to_end/workflow/reference_include_flow.py

```diff
@@ -0,0 +1,163 @@
+        self.collab_attr_dict_two = {key: key for key in range(5)}
+
+        TestFlowReferenceWithInclude.step_two_collab_attrs.append(self)
+
+        if (
+            len(TestFlowReferenceWithInclude.step_two_collab_attrs)
+            >= MIN_COLLECTION_COUNT
+        ):
+            collab_attr_list = filter_attrs(inspect.getmembers(self))
+            matched_ref_dict = find_matched_references(
+                collab_attr_list,
+                TestFlowReferenceWithInclude.step_two_collab_attrs,
+            )
+            validate_references(matched_ref_dict)
+
+        self.next(self.join, include=["collab_attr_dict_two"])
+
+    @aggregator
+    def join(self, inputs):
+        all_attr_list = filter_attrs(inspect.getmembers(inputs[0]))
+
+        matched_ref_dict = find_matched_references(all_attr_list, inputs)
+        validate_references(matched_ref_dict)
+        all_shared_attr = ""
+        log.info("Reference test summary:")
+        for val in matched_ref_dict.values():
+            all_shared_attr = all_shared_attr + ",".join(val)
+        if all_shared_attr:
+            log.error(f"Test case failed for {all_shared_attr}")
+        else:
+            log.info("Test case passed for all the attributes.")
+        self.next(self.end)
+
+    @aggregator
+    def end(self):
+        log.info("Testing FederatedFlow - Ending test for validating the references.")
+        TestFlowReferenceWithInclude.step_one_collab_attrs = []
+        TestFlowReferenceWithInclude.step_two_collab_attrs = []
+
+
+def filter_attrs(attr_list):
+    valid_attrs = []
+    reserved_words = ["next", "runtime", "execute_next"]
+    for attr in attr_list:
+        if (
+            not attr[0].startswith("_")
+            and attr[0] not in reserved_words
+            and not hasattr(TestFlowReferenceWithInclude, attr[0])
+        ):
+            if not isinstance(attr[1], MethodType):
+                valid_attrs.append(attr[0])
+    return valid_attrs
+
+
+def find_matched_references(collab_attr_list, all_collaborators):
+    matched_ref_dict = {}
+    for i in range(len(all_collaborators)):
+        matched_ref_dict[all_collaborators[i].input] = []
+
+    for attr_name in collab_attr_list:
+        for i, curr_collab in enumerate(all_collaborators):
+            for next_collab in all_collaborators[i + 1:]:
+                if hasattr(curr_collab, attr_name) and hasattr(next_collab, attr_name):
+                    if getattr(curr_collab, attr_name) is getattr(
+                        next_collab, attr_name
+                    ):
+                        matched_ref_dict[curr_collab.input].append(attr_name)
+                        log.error(
+                            f"Reference test failed - {curr_collab.input} sharing same "
+                            + f"{attr_name} reference with {next_collab.input}"
+                        )
+
+    return matched_ref_dict
+
+
+def validate_references(matched_ref_dict):
```

Comment:
Please add doc-string. The code logic is not straightaway understood, so provide an example if possible.

Reply:
done